aesara/tensor/basic.py

```diff
@@ -250,338 +244,88 @@ def constant(x, name=None, ndim=None, dtype=None):
-            data,
-        )
-
-
-_scalar_constant_value_elemwise_ops = (
-    aes.Cast,
-    aes.Switch,
-    aes.NEQ,
-    aes.EQ,
-    aes.LT,
-    aes.GT,
-    aes.LE,
-    aes.GE,
-    aes.Sub,
-    aes.Add,
-    aes.Mod,
-    aes.Mul,
-    aes.IntDiv,
-    aes.TrueDiv,
-    aes.ScalarMinimum,
-    aes.ScalarMaximum,
-)
-
-
-def get_scalar_constant_value(
-    orig_v, elemwise=True, only_process_constants=False, max_recur=10
-):
-    """Return the constant scalar(0-D) value underlying variable `v`.
-
-    If `v` is the output of dimshuffles, fills, allocs, rebroadcasts,
-    cast, OutputGuard, DeepCopyOp, ScalarFromTensor, ScalarOp, Elemwise
-    and some pattern with Subtensor, this function digs through them.
+def get_scalar_constant_value(v: Variable, only_process_constants=False, max_recur=10):
+    r"""Return the constant scalar value for `v`.
 
-    If `v` is not some view of constant scalar data, then raise a
-    NotScalarConstantError.
+    This will perform canonicalization and constant folding on a clone of `v`.
 
     Parameters
     ----------
-    elemwise : bool
-        If False, we won't try to go into elemwise. So this call is faster.
-        But we still investigate in Second Elemwise (as this is a substitute
-        for Alloc)
     only_process_constants : bool
         If True, we only attempt to obtain the value of `orig_v` if it's
         directly constant and don't try to dig through dimshuffles, fills,
         allocs, and other to figure out its value.
-    max_recur : int
-        The maximum number of recursion.
+
+
+    Raises
+    ------
+    If `v` is not some view of constant scalar data, then raise a
+    `NotScalarConstantError`.
 
     Notes
     -----
         There may be another function similar to this one in the code,
         but I'm not sure where it is.
 
     """
-    v = orig_v
-    while True:
-        if v is None:
-            # None is not a scalar (and many uses of this function seem
-            # to depend on passing it None)
+    if isinstance(v, (np.integer, int, float)):
+        return np.asarray(v)
+
+    if getattr(v, "ndim", None) != 0:
+        raise NotScalarConstantError()
+
+    if isinstance(v, np.ndarray):
+        try:
+            return np.array(v.item())
+        except ValueError:
             raise NotScalarConstantError()
 
-        if isinstance(v, (np.integer, int, float)):
-            return np.asarray(v)
+    if isinstance(v, Constant):
+        # TODO: Get rid of this hack.
+        if getattr(v.tag, "unique_value", None) is not None:
+            data = v.tag.unique_value
+        else:
+            data = v.data
 
-        if isinstance(v, np.ndarray):
-            return numpy_scalar(v).copy()
+        if isinstance(data, np.ndarray):
+            try:
+                return np.array(data.item(), dtype=v.dtype)
+            except ValueError:
+                raise NotScalarConstantError()
+        else:
+            return data
```

Comment:
Is this logic repetition intentional ?
L283-296 and L298-310 seems to be doing the same thing.